src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpConnectionFactory.cs

```diff
@@ -79,11 +79,34 @@ public virtual async ValueTask<Connection> EstablishConnectionAsync(HttpRequestM
                 socket.NoDelay = true;
                 return new SocketConnection(socket);
             }
+            catch (SocketException socketException)
+            {
+                socket.Dispose();
+                throw MapSocketException(socketException);
+            }
             catch
             {
                 socket.Dispose();
                 throw;
             }
         }
+
+        internal static NetworkException MapSocketException(SocketException socketException)
+        {
+            NetworkError error = socketException.SocketErrorCode switch
+            {
+                SocketError.AddressAlreadyInUse => NetworkError.AddressInUse,
+                SocketError.AddressFamilyNotSupported => NetworkError.InvalidAddress,
+                SocketError.AddressNotAvailable => NetworkError.InvalidAddress,
+                SocketError.HostNotFound => NetworkError.HostNotFound,
+                SocketError.ConnectionRefused => NetworkError.ConnectionRefused,
+                SocketError.OperationAborted => NetworkError.OperationAborted,
+                SocketError.ConnectionAborted => NetworkError.ConnectionAborted,
+                SocketError.ConnectionReset => NetworkError.ConnectionReset,
+                _ => NetworkError.Unknown
+            };
```

Comment:
It probably doesn't matter, but code size can be reduced by using a switch statement rather than a switch expression, by having the `NetworkError.InvalidAddress` cases fall through to a single branch.

Reply:
I'm kind of surprised that switch expressions don't support this sort of thing...